cpu/cc2538/dev/spi.c

```diff
@@ -41,102 +41,337 @@
+/* Here we check that either all or none of the ports are defined. As
+   we did already check that both ports + pins are either defined or
+   not for every pin, this means that we can check for an incomplete
+   configuration by only looking at the port defines */
+#if (SPI1_CLK_PORT < 0 && SPI1_TX_PORT < 0 && SPI1_RX_PORT >= 0) || \
+  (SPI1_CLK_PORT < 0 && SPI1_TX_PORT >= 0 && SPI1_RX_PORT < 0) || \
+  (SPI1_CLK_PORT < 0 && SPI1_TX_PORT >= 0 && SPI1_RX_PORT >= 0) || \
+  (SPI1_CLK_PORT >= 0 && SPI1_TX_PORT < 0 && SPI1_RX_PORT < 0) || \
+  (SPI1_CLK_PORT >= 0 && SPI1_TX_PORT < 0 && SPI1_RX_PORT >= 0) || \
+  (SPI1_CLK_PORT >= 0 && SPI1_TX_PORT >= 0 && SPI1_RX_PORT < 0)
+#error "SPI1 port / pin definition incomplete"
+#endif
+/*---------------------------------------------------------------------------*/
+typedef struct {
+  int8_t port;
+  int8_t pin;
+} spi_pad_t;
+typedef struct {
+  uint32_t base;
+  uint32_t ioc_ssirxd_ssi;
+  uint32_t ioc_pxx_sel_ssi_clkout;
+  uint32_t ioc_pxx_sel_ssi_txd;
+  spi_pad_t clk;
+  spi_pad_t tx;
+  spi_pad_t rx;
+} spi_cfg_t;
+/*---------------------------------------------------------------------------*/
+static const spi_cfg_t spi_cfg[SSI_INSTANCE_COUNT] = {
+  {
+    .base = SSI0_BASE,
+    .ioc_ssirxd_ssi = IOC_SSIRXD_SSI0,
+    .ioc_pxx_sel_ssi_clkout = IOC_PXX_SEL_SSI0_CLKOUT,
+    .ioc_pxx_sel_ssi_txd = IOC_PXX_SEL_SSI0_TXD,
+    .clk = { SPI0_CLK_PORT, SPI0_CLK_PIN },
+    .tx = { SPI0_TX_PORT, SPI0_TX_PIN },
+    .rx = { SPI0_RX_PORT, SPI0_RX_PIN }
+  }, {
+    .base = SSI1_BASE,
+    .ioc_ssirxd_ssi = IOC_SSIRXD_SSI1,
+    .ioc_pxx_sel_ssi_clkout = IOC_PXX_SEL_SSI1_CLKOUT,
+    .ioc_pxx_sel_ssi_txd = IOC_PXX_SEL_SSI1_TXD,
+    .clk = { SPI1_CLK_PORT, SPI1_CLK_PIN },
+    .tx = { SPI1_TX_PORT, SPI1_TX_PIN },
+    .rx = { SPI1_RX_PORT, SPI1_RX_PIN }
+  }
+};
+/*---------------------------------------------------------------------------*/
+/* Deprecated function calls provided for compatibility reasons */
 void
 spi_init(void)
 {
-  spi_enable();
+  spix_init(SPI_DEFAULT_INSTANCE);
+}
+void
+spi_enable(void)
+{
+  spix_enable(SPI_DEFAULT_INSTANCE);
+}
+void
+spi_disable(void)
+{
+  spix_disable(SPI_DEFAULT_INSTANCE);
+}
+void
+spi_set_mode(uint32_t frame_format, uint32_t clock_polarity,
+             uint32_t clock_phase, uint32_t data_size)
+{
+  spix_set_mode(SPI_DEFAULT_INSTANCE, frame_format, clock_polarity,
+                clock_phase, data_size);
+}
+void
+spi_cs_init(uint8_t port, uint8_t pin)
+{
+  spix_cs_init(port, pin);
+}
+/*---------------------------------------------------------------------------*/
+void
+spix_init(uint8_t instance)
+{
+  const spi_cfg_t *cfg;
+
+  if(instance >= SSI_INSTANCE_COUNT) {
+    return;
+  }
+
+  spix_enable(instance);
+
+  cfg = &spi_cfg[instance];
+
+  if(cfg->clk.port < 0) {
+    /* Port / pin configuration invalid. We checked for completeness
+       above. If clk.port is < 0, this means that all other defines are
+       < 0 as well */
+    return;
+  }
```

Comment:
Preliminary validity checks must be performed before actually doing anything with the hardware, i.e. before calling `spix_enable()`.